File: 比赛/2017/1.24/fast.pas
var
    a:array[1..20]of char;
	data:array[1..20]of integer;
	n,i,h,ans,len:integer;
procedure merge;
begin
	data[h-1]:=data[h-1]+data[h];
	dec(h);
	ans:=ans+1;
end;
begin
	assign(input,'fast.in');
	assign(output,'fast.out');
	reset(input);
	rewrite(output);
	readln(len);
    for i:=1 to len do
        read(a[i]);
    for i:=1 to len do
        n:=n*2+ord(a[i])-ord('0');
	h:=1;
	data[h]:=1;
	ans:=0;
	for i:=2 to n do
	begin
		inc(h);
		data[h]:=1;
		while (h>1)and(data[h]=data[h-1]) do
		merge;
	end;
	writeln(ans);
	close(input);
	close(output);
end.

File: 比赛/2017/1.24/fast.cpp
#include<iostream>
#include<cstdio>
#include<cstring>
#include<algorithm>
using namespace std;
char a[20];
int data[20];
int n,i,h,ans,len;
void merge()
{
	data[h-1]=data[h-1]+data[h];
	h--;
	ans++;
}
int main()
{
	scanf("%d",&len);
	scanf("%s",&a);
	for(i=0;i<len;i++)
		n=n*2+a[i]-'0';
	h=1;
	data[h]=1;
	ans=0;
	for(i=2;i<=n;i++)
	{
		h++;
		data[h]=1;
		while(h>1&&data[h]==data[h-1])
			merge();
	}
	printf("%d",ans);
	return 0;
}
